### PR TITLE
chore: updates for okta-auth-js@3.0 - OKTA-276557

### PR DESCRIPTION
**Pull request by aarongranick-okta**
- Brings in [changes from AuthjS 3.0](https://github.com/okta/okta-auth-js/blob/master/CHANGELOG.md#300)
- PKCE is true by default
- PKCE will use responseMode=query by default
- jQuery removed from unit tests
- Q removed from AuthJS, replaced with standard promise
- showSignInToGetTokens changed:
  - getAccessToken is now `true` by default
  - `authorizationServerId` option has been removed from To specify a custom authorization server, use `authParams.issuer`. The issuer should be specified as a full URI, not just the server ID.
- minimum node version is 10

### CODE REVIEW

File: src/util/BaseLoginRouter.js [hunk 1]
```diff
@@ -31,7 +32,7 @@ define([
   'util/Bundles',
   'util/Logger'
 ],
-function (Okta, BrowserFeatures, Settings,
+function (Q, Okta, BrowserFeatures, Settings,
```
Comment (nikhilvenkatraman-okta) [inline src/util/BaseLoginRouter.js]: Curious why this change is preferred over what we had before.

File: src/util/OAuth2Util.js [hunk 2]
```diff
@@ -48,54 +48,50 @@ define(['okta', './Enums', './Errors', './Util'], function (Okta, Enums, Errors,
 
     var authClient = settings.getAuthClient(),
         options = settings.toJSON({ verbose: true }),
-        oauthParams = {},
-        extraOptions = {};
+        getTokenOptions = {};
```
Comment (nikhilvenkatraman-okta) [inline src/util/OAuth2Util.js]: nit: is it better to call this just tokenOptions? `getTokenOptions ` sounds more like a fn name.

File: test/e2e/templates/done.tpl [hunk 3]
```diff
@@ -2,27 +2,42 @@
 // OIDC Redirect Flow - this is the page that is redirected to with
 // tokens in the parameters
 
+// PKCE cannot be enabled because the test app is a "web" type. OKTA-246000 
+var pkce = false;
```
Comment (nikhilvenkatraman-okta) [inline test/e2e/templates/done.tpl]: Is this why we are setting `pkce: false` to a bunch of tests in the widget ?

File: test/unit/helpers/mocks/Util.js [hunk 4]
```diff
@@ -59,75 +69,105 @@ function (Okta, Q, Duo, keys, wellKnown, wellKnownSharedResource) {
   };
 
   fn.mockAjax = function (responses) {
+    jasmine.Ajax.install();
```
Comment (nikhilvenkatraman-okta) [inline test/unit/helpers/mocks/Util.js]: 👏

File: test/unit/spec/ConsentRequired_spec.js [hunk 5]
```diff
@@ -155,14 +155,14 @@ function (Okta, OktaAuth, LoginUtil, Util, ConsentRequiredForm, Expect, Router,
...
+          Util.resetAjaxRequests();
           test.setNextResponse(resSuccess);
           test.form.consentButton().click();
-          return Expect.waitForSpyCall($.ajax);
+          return Expect.waitForAjaxRequest();
         })
           .then(function () {
-            expect($.ajax.calls.count()).toBe(1);
-            Expect.isJsonPost($.ajax.calls.argsFor(0), {
+            expect(Util.numAjaxRequests()).toBe(1);
```
Comment (nikhilvenkatraman-okta) [inline test/unit/spec/ConsentRequired_spec.js]: 👍

File: src/util/OAuth2Util.js (hunk 2 above)
Comment (aarongranick-okta) [reply src/util/OAuth2Util.js]: Or maybe `authParams`? We use this name in other places to refer to the config passed to the authClient. The options we are building here correspond with this documentation: https://github.com/okta/okta-auth-js#authorize-options Maybe `authOptions`?

File: src/util/BaseLoginRouter.js (hunk 1 above)
Comment (aarongranick-okta) [reply src/util/BaseLoginRouter.js]: I don't remember exactly. My guess is it has to do with unhandled promise rejection? I'll change it back and see what effect it has on the tests.

File: test/e2e/templates/done.tpl (hunk 3 above)
Comment (aarongranick-okta) [reply test/e2e/templates/done.tpl]: In the E2E tests doing OIDC flow, yes. They were built using implicit flow. I have a [task](https://oktainc.atlassian.net/browse/OKTA-246000) to add a SPA client ID to the test config. I would probably want to add this to the 4.0 branch before we merge to master.

File: src/util/OAuth2Util.js (hunk 2 above)
Comment (nikhilvenkatraman-okta) [reply src/util/OAuth2Util.js]: either works

File: src/util/BaseLoginRouter.js (hunk 1 above)
Comment (aarongranick-okta) [reply src/util/BaseLoginRouter.js]: There is a mixed usage of Q and standard promise. I am adding another commit which changes `fail` to `catch` in these cases so we can accept either.

File: src/views/shared/Header.js [hunk 6]
```diff
@@ -139,7 +139,7 @@ function (Okta, Animations, LoadingBeacon) {
             removeBeacon(self);
             addBeacon(self, NextBeacon, selector, options);
           })
-          .done();
+          .done(); // TODO: can this be removed if fadeOut returns standard ES6 Promise?
```
Comment (nikhilvenkatraman-okta) [inline src/views/shared/Header.js]: Should we create a JIRA to refactor this later i.e use standard ES6 Promise ?

**Comment by jhe-okta** [comment]
I will leave it to @nikhilvenkatraman-okta since he has the context.
